### ISSUE TITLE
No saved models

### ISSUE THREAD

**Issue by Delev1n** [sarulab-speech/Sidon#8]
When I ran the first stage of training on my custom data, even tho the train ran and finished, no checkpoints were saved. I added this to ```default.yaml```:

```
- _target_: lightning.pytorch.callbacks.ModelCheckpoint
- dirpath: ${hydra:runtime.output_dir}/checkpoints
- filename: "{epoch}-{step}"
- save_last: true
- save_top_k: -1
- every_n_train_steps: 2000
- save_on_train_epoch_end: false
```

Now some checkpoints were saved, but in the wrong format. Maybe I am doing something wrong?

**Comment by Wataru-Nakata** [comment]
I'm not really sure wha'ts happening but One possibility is that the training data is too small to have a saved checkpoint. Can you add samples in a dataset?

Also, having wrong format checkpoint is indeed weird. Can you tell me what kind of error you encountered?

**Comment by Delev1n** [comment]
I have ~90 hours of already paired data that I converted to the needed format using the following script:

```
import os
import math
import pickle
import random
from pathlib import Path

import pandas as pd
import torch
import torchaudio
import webdataset as wds

from transformers import AutoFeatureExtractor

fe = AutoFeatureExtractor.from_pretrained("facebook/w2v-bert-2.0")

STAGE = 1
PHASE = "train"

CSV_PATH = f"{PHASE}.csv"
OUT_DIR = Path(f"new_data/stage_1/wds_out_{PHASE}")
OUT_DIR.mkdir(parents=True, exist_ok=True)

NOISY_COL = "noisy_path"
CLEAN_COL = "enhanced_path"

WAV_SR = 16_000
SSL_SR = 16_000

SEG_SECONDS = 6.0 
SEG_SAMPLES = int(WAV_SR * SEG_SECONDS)

SAMPLES_PER_SHARD = 1000
SEED = 0
random.seed(SEED)

DEVICE_FOR_SSL_FEATS = "cpu"


def load_mono(path):
    wav, sr = torchaudio.load(path)
    if wav.ndim == 2 and wav.shape[0] > 1:
        wav = wav.mean(dim=0, keepdim=True)
    return wav.squeeze(0), sr


def resample_if_needed(wav, sr, target_sr):
    if sr == target_sr:
        return wav
    return torchaudio.functional.resample(wav, sr, target_sr)


def pair_peak_only(noisy, clean, eps=1e-9, peak_target=0.95):
    peak = noisy.abs().max().clamp_min(eps)
    scale = peak_target / peak
    return noisy * scale, clean * scale


def split_into_segments_pair(noisy, clean, seg_len, hop_len):
    T = min(noisy.numel(), clean.numel())
    noisy = noisy[:T]
    clean = clean[:T]

    start = 0
    idx = 0

    while start < T:
        end = start + seg_len

        if end <= T:
            noisy_seg = noisy[start:end]
            clean_seg = clean[start:end]
        else:
            noisy_seg = torch.nn.functional.pad(noisy[start:T], (0, end - T))
            clean_seg = torch.nn.functional.pad(clean[start:T], (0, end - T))

        yield idx, noisy_seg, clean_seg

        idx += 1
        start += hop_len


def make_ssl_inputs(wav, sr):
    if sr != SSL_SR:
        wav = torchaudio.functional.resample(wav, sr, SSL_SR)

    wav_np = wav.detach().cpu().float().numpy()
    feats = fe(wav_np, sampling_rate=SSL_SR, return_tensors="pt")

    out = {}
    for k, v in feats.items():
        out[k] = (v.squeeze(0).cpu(),)
    return out


def open_sink(sid):
    shard_path = OUT_DIR / f"shard-{sid:05d}.tar"
    return wds.TarWriter(str(shard_path))


df = pd.read_csv(CSV_PATH)

print(f"Writing {len(df)} rows into shards of {SAMPLES_PER_SHARD} samples...")

sample_count = 0
shard_id = 0

sink = open_sink(shard_id)

for row_idx in range(len(df)):
    noisy_path = str(df.iloc[row_idx][NOISY_COL])
    clean_path = str(df.iloc[row_idx][CLEAN_COL])

    noisy_wav, noisy_sr = load_mono(noisy_path)
    clean_wav, clean_sr = load_mono(clean_path)

    noisy_wav = resample_if_needed(noisy_wav, noisy_sr, WAV_SR)
    clean_wav = resample_if_needed(clean_wav, clean_sr, WAV_SR)

    noisy_wav, clean_wav = pair_peak_only(noisy_wav, clean_wav)

    for seg_idx, noisy_seg, clean_seg in split_into_segments_pair(
        noisy_wav, clean_wav, SEG_SAMPLES, SEG_SAMPLES
    ):
        if sample_count > 0 and sample_count % SAMPLES_PER_SHARD == 0:
            sink.close()
            shard_id += 1
            sink = open_sink(shard_id)

        noisy_ssl = make_ssl_inputs(noisy_seg, WAV_SR)
        clean_ssl = make_ssl_inputs(clean_seg, WAV_SR)

        key = f"utt_{row_idx:09d}_{seg_idx:03d}"
        sample = {
            "__key__": key,
            "input_wav.pth": clean_seg.contiguous(),
            "noisy_input_wav.pth": noisy_seg.contiguous(),
            "sr.index": WAV_SR,
            "ssl_inputs.pickle": clean_ssl,
            "noisy_ssl_inputs.pickle": noisy_ssl,
        }
        sink.write(sample)
        sample_count += 1

sink.close()
print(f"Done. Wrote {sample_count} samples into {shard_id+1} shards.")
```

I got around 21 shards with 1000 samples inside. Trained it for 32K steps and got nothing in the output, even tho the tensorboard showed how the train loss was changing.

**Comment by Wataru-Nakata** [comment]
Can you try calling trainer.save_checkpoint manually in the training script to see if that works or not?

**Comment by Delev1n** [comment]
It indeed saved, but both the teacher and student models in one checkpoint, so it's still not usable in the inference script by default.

**Comment by Wataru-Nakata** [comment]
To create inference weight, I performed torchscript jit compilation for the best performance. You can also do the same if you only need feature predictor and vocoder.
